### PR TITLE
[vectorized](pipeline) support union node operator

### PR DESCRIPTION
**Pull request by zhangstar333**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.08 seconds
 load time: 455 seconds
 storage size: 17123356932 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221213061759_clickbench_pr_62410.html

File: be/src/pipeline/exec/data_queue.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include <mutex>
+#include <vector>
+
+namespace doris {
+namespace vectorized {
+class Block;
+}
+namespace pipeline {
+
+class DataQueue {
```
Comment (HappenLee) [inline be/src/pipeline/exec/data_queue.h:31]: maybe we should replace `AggContext` with `DataQueue`

File: be/src/pipeline/exec/union_sink_operator.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    auto output_block = std::make_unique<vectorized::Block>();
+    if (_cur_child_id < _node->get_first_materialized_child_idx()) { //pass_through
+        output_block->swap(*in_block);
+    } else if (_node->get_first_materialized_child_idx() !=
+                       _node->children_count() && //need materialized
+               _cur_child_id < _node->children_count()) {
+        this->_node->materialize_child_block(state, _cur_child_id, in_block, output_block.get());
+    } else {
+        LOG(WARNING) << "maybe can't reach here, execute const expr: " << _cur_child_id << " "
```
Comment (HappenLee) [inline be/src/pipeline/exec/union_sink_operator.cpp:49]: return Error Status

File: be/src/vec/exec/vunion_node.cpp [hunk 3]
```diff
@@ -220,6 +221,21 @@ Status VUnionNode::get_next_const(RuntimeState* state, Block* block) {
     return Status::OK();
 }
 
+//for pipeline operator
+Status VUnionNode::materialize_child_block(RuntimeState* state, int child_id,
+                                           vectorized::Block* input_block,
+                                           vectorized::Block* output_block) {
+    DCHECK_LT(child_id, _children.size());
+    DCHECK(!is_child_passthrough(child_id));
+    MutableBlock mblock = MutableBlock(
```
Comment (HappenLee) [inline be/src/vec/exec/vunion_node.cpp:230]: rethink the mem reuse and try to merge the block to make sure output_block have `batch_size`.

Comment (HappenLee) [inline be/src/vec/exec/vunion_node.cpp:230]: rethink the mem reuse and try to merge the block to make sure output_block have `batch_size`.

File: be/src/pipeline/exec/data_queue.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+#include "vec/core/block.h"
+namespace doris {
+namespace vectorized {
+class Block;
+}
+namespace pipeline {
+
+class DataQueue {
+public:
+    //always one is enough, but in union node it's has more childen
```
Comment (HappenLee) [inline be/src/pipeline/exec/data_queue.h:34]: children

File: be/src/pipeline/exec/data_queue.cpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    return false;
+}
+
+//the _flag_queue_idx indicate which queue has data, and in check can_read
+//will be set idx in remaining_has_data function
+Status DataQueue::get_block_from_queue(std::unique_ptr<vectorized::Block>* output_block,
+                                       int* child_idx) {
+    if (_is_canceled[_flag_queue_idx]) {
+        return Status::InternalError("AggContext canceled");
```
Comment (HappenLee) [inline be/src/pipeline/exec/data_queue.cpp:100]: ```AggContext canceled``` ?

Comment (HappenLee) [inline be/src/pipeline/exec/data_queue.cpp:100]: ```AggContext canceled``` ?

File: be/src/pipeline/exec/data_queue.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    if (_cur_blocks_nums_in_queue[_flag_queue_idx] > 0) {
```
Comment (HappenLee) [inline be/src/pipeline/exec/data_queue.cpp:103]: danger !!! two thread call function will coredump

File: be/src/pipeline/exec/union_sink_operator.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    } else {
+        return Status::InternalError("maybe can't reach here, execute const expr: {}, {}, {}",
+                                     _cur_child_id, _node->get_first_materialized_child_idx(),
+                                     _node->children_count());
+    }
+
+    if (UNLIKELY(source_state == SourceState::FINISHED)) {
+        //if _cur_child_id eos, need check to push block
+        if (_output_block) {
+            _data_queue->push_block(std::move(_output_block), _cur_child_id);
```
Comment (HappenLee) [inline be/src/pipeline/exec/union_sink_operator.cpp:63]: check output_block->rows() == 0

File: be/src/pipeline/exec/union_sink_operator.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    if (_output_block == nullptr) {
+        _output_block = _data_queue->get_free_block(_cur_child_id);
+    }
+
+    if (_cur_child_id < _node->get_first_materialized_child_idx()) { //pass_through
+        if (in_block->rows() > 0) {
+            _output_block->swap(*in_block);
+            _data_queue->push_block(std::move(_output_block), _cur_child_id);
+            _output_block.reset(nullptr);
```
Comment (HappenLee) [inline be/src/pipeline/exec/union_sink_operator.cpp:49]: do not need call `reset`

File: be/src/pipeline/exec/union_sink_operator.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        //if _cur_child_id eos, need check to push block
+        if (_output_block) {
+            _data_queue->push_block(std::move(_output_block), _cur_child_id);
+            _output_block.reset(nullptr);
+        }
+        _data_queue->set_finish(_cur_child_id);
+        return Status::OK();
+    }
+    // not eos and block rows is enough to output,so push block
+    if (_output_block && (_output_block->rows() > state->batch_size())) {
```
Comment (HappenLee) [inline be/src/pipeline/exec/union_sink_operator.cpp:70]: >=

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**